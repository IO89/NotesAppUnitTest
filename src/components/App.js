import React, { Component } from "react";
import { Button } from "react-bootstrap";
import Note from "./Note";

class App extends Component {
  constructor() {
    super();

    this.state = { notes: [] };
  }

  addNote = () => {
    const { notes } = this.state;
    const ids = this.state.notes.map(note => note.id);
    const max_id = ids.length > 0 ? Math.max(...ids) : 0;

    notes.push({ id: max_id + 1 });

    this.setState({ notes });
  };

  removeNote = id => {
    const notes = this.state.notes.filter(note => note.id !== id);
    this.setState({ notes });
  };

  render() {
    return (
      <div>
        <h2>Unit Test App</h2>
        <div className="notes-list">
          {this.state.notes.map(note => {
            return (
              <Note key={note.id} 
              note={note} 
              removeNote={this.removeNote}
               />
            );
          })}
        </div>
        <Button className="btn-add" onClick={this.addNote}>
          Add Note
        </Button>
      </div>
    );
  }
}

export default App;
