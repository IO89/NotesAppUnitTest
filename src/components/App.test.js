import React from "react";
import { shallow } from "enzyme";
import App from "./App";

describe("App", () => {
  const app = shallow(<App />);

  it("renders correctly", () => {
    expect(app).toMatchSnapshot();
  });

  it("initialize the `state` with an empty list", () => {
    expect(app.state().notes).toEqual([]);
  });

  describe("When clicking `add button`", () => {
    const id = 1;

    beforeEach(() => {
      app.find(".btn-add").simulate("click");
    });

    afterEach(() => {
      app.setState({ notes: [] });
    });

    it("adds new note to `state`", () => {
      expect(app.state().notes).toEqual([{ id }]);
    });

    it("adds new note to rendered list", () => {
      expect(app.find(".notes-list").children().length).toEqual(1);
    });

    it("creates a Note component", () => {
      expect(app.find("Note").exists()).toBe(true);
    });

    describe("When person want to remove added note", () => {
      beforeEach(() => {
        app.instance().removeNote(id);
      });

      it("remove note from `state`", () => {
        expect(app.state().notes).toEqual([]);
      });
    });
  });
});
