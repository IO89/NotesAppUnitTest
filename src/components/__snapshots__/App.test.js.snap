// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Unit Test App
</h2>,
        <div
          className="notes-list"
>
          
</div>,
        <Button
          active={false}
          block={false}
          bsClass="btn"
          bsStyle="default"
          className="btn-add"
          disabled={false}
          onClick={[Function]}
>
          Add Note
</Button>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Unit Test App",
        },
        "ref": null,
        "rendered": "Unit Test App",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "notes-list",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": false,
          "block": false,
          "bsClass": "btn",
          "bsStyle": "default",
          "children": "Add Note",
          "className": "btn-add",
          "disabled": false,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Add Note",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Unit Test App
</h2>,
          <div
            className="notes-list"
>
            
</div>,
          <Button
            active={false}
            block={false}
            bsClass="btn"
            bsStyle="default"
            className="btn-add"
            disabled={false}
            onClick={[Function]}
>
            Add Note
</Button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Unit Test App",
          },
          "ref": null,
          "rendered": "Unit Test App",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "notes-list",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": false,
            "block": false,
            "bsClass": "btn",
            "bsStyle": "default",
            "children": "Add Note",
            "className": "btn-add",
            "disabled": false,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Add Note",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "disableLifecycleMethods": true,
    "hydrateIn": undefined,
  },
}
`;
